packages/path-provider/example/lib/main.dart

```diff
@@ -0,0 +1,132 @@
+
+  void _requestTempDirectory() {
+    setState(() {
+      _tempDirectory = getTemporaryDirectory();
+    });
+  }
+
+  Widget _buildTempDirectory(
+      BuildContext context, AsyncSnapshot<Directory> snapshot) {
+    if (snapshot.hasError) {
+      return new Padding(
+          padding: const EdgeInsets.all(16.0),
+          child: new Text('Error: ${snapshot.error}'));
+    } else if (snapshot.hasData) {
+      return new Padding(
+          padding: const EdgeInsets.all(16.0),
+          child: new Text('path: ${snapshot.data.path}'));
+    } else {
+      return new Padding(
+          padding: const EdgeInsets.all(16.0), child: new Text(''));
+    }
+  }
+
+  void _requestAppDocumentsDirectory() {
+    setState(() {
+      _appDocumentsDirectory = getApplicationDocumentsDirectory();
+    });
+  }
+
+  Widget _buildAppDocumentsDirectory(
+      BuildContext context, AsyncSnapshot<Directory> snapshot) {
+    if (snapshot.hasError) {
+      return new Padding(
+          padding: const EdgeInsets.all(16.0),
+          child: new Text('Error: ${snapshot.error}'));
+    } else if (snapshot.hasData) {
+      return new Padding(
+          padding: const EdgeInsets.all(16.0),
+          child: new Text('path: ${snapshot.data.path}'));
+    } else {
+      return new Padding(
+          padding: const EdgeInsets.all(16.0), child: new Text(''));
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return new Scaffold(
+      appBar: new AppBar(
+        title: new Text(widget.title),
+      ),
+      body: new Center(
+        child: new Column(
+          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
+          children: <Widget>[
+            new Column(
+              mainAxisAlignment: MainAxisAlignment.start,
+              children: <Widget>[
+                new Padding(
+                  padding: const EdgeInsets.all(16.0),
+                  child: new RaisedButton(
+                    child: const Text('Get Temporary Directory'),
+                    onPressed: _requestTempDirectory,
+                  ),
+                ),
+              ],
+            ),
+            new Expanded(
+              child: new FutureBuilder<Directory>(
+                  future: _tempDirectory, builder: _buildTempDirectory),
+            ),
+            new Column(
+              mainAxisAlignment: MainAxisAlignment.start,
+              children: <Widget>[
+                new Padding(
+                  padding: const EdgeInsets.all(16.0),
+                  child: new RaisedButton(
+                    child: const Text('Get Application Documents Directory'),
+                    onPressed: _requestAppDocumentsDirectory,
+                  ),
+                ),
+              ],
+            ),
+            new Expanded(
+              child: new FutureBuilder<Directory>(
+                  future: _appDocumentsDirectory,
+                  builder: _buildAppDocumentsDirectory),
+            ),
+          ],
+        ),
```

Comment:
Again, there is duplication above with the same tree structure twice. Wonder if it would be worth it to refactor. Up to you.